### PR TITLE
Contributing file updates, remove old development file

### PR DESCRIPTION
**Pull request by lornajane**
A pretty major overhaul of the contributing file. The main aims are to fill in gaps, capture informal conventions, and try to structure the information more for newcomers as well as project regulars.

 A huge number of "we should write this down" issues are (at least partially) addressed in this update:

- #3518 
- #3715 
- #3716 
- #3802 
- #3523 
- #3524 
- #3677 
- #3847 
- #3848 
- #3581 

I think there's a lot here that probably isn't perfect, but let's try to merge when it's good enough to be considered an improvement on our current file, even if there are followup actions to take?

### CODE REVIEW

**Review by ralfhandl — CHANGES_REQUESTED** [review]
Minor editorial suggestions, formatting, and corrected links

File: CONTRIBUTING.md
Comment (lornajane) [reply CONTRIBUTING.md:12]: I consider this to be key information but I'll expand it to make it clear that this file is only on branches.

File: CONTRIBUTING.md
Comment (lornajane) [reply CONTRIBUTING.md:26]: This is good information but I don't think file location information is best placed here, I'll move it.

File: CONTRIBUTING.md
Comment (ralfhandl) [inline CONTRIBUTING.md:137]:
```suggestion
> Check the [Appendix: Branch Strategy](#appendix-branch-strategy) for more information about the updated branching strategy.
```

File: CONTRIBUTING.md
Comment (ralfhandl) [inline CONTRIBUTING.md:257]:
```suggestion
   Use the [template in the proposals directory](https://github.com/OAI/OpenAPI-Specification/tree/main/proposals/2019-01-01-Proposal-Template.md), copy it to a new file, and complete it.
```

**Review by ralfhandl — CHANGES_REQUESTED** [review]
+1, only broken links and minor nits

File: CONTRIBUTING.md
Comment (handrews) [inline CONTRIBUTING.md:13]: Can we use 3.1 as the example here?  The next 3.2 release is 3.2.0, which we would not normally refer to as a "patch release"?  Or just change "patch release" to "release."

File: CONTRIBUTING.md
Comment (handrews) [inline CONTRIBUTING.md:121]: I made a point to put PRs in milestones, and then when we released 3.0.4 and 3.1.1 I realized that you can look them up by target branch, which for us corresponds directly to releases.  So maybe we don't need to put milestones on PRs after all?  They should definitely go on issues, though.

File: CONTRIBUTING.md
Comment (handrews) [inline CONTRIBUTING.md:120]: Specifically, use the oldest-release-line milestone that is relevant if we are working on multiple releases at once.

File: CONTRIBUTING.md
Comment (handrews) [inline CONTRIBUTING.md:142]: Should GitHub discussions be mentioned there?  Earlier parts of this document imply that big things go to discussions, actionable thing go to issues, and the formal proposal process seems to be a way to bridge the gap by formalizing a discussion that might not be ready for individual issues yet.  At least that's what I get from it?

File: CONTRIBUTING.md
Comment (handrews) [inline CONTRIBUTING.md:152]: earliest relevant active

File: CONTRIBUTING.md
Comment (handrews) [inline CONTRIBUTING.md:161]: There might be things we want to start on `dev`, but I'm just saying that to make a note of it- I don't think there is anything to say about it yet.  I think we just need to figure out a bit more what counts as infrastructure that is not needed on `main` (and shouldn't clutter it up) vs information that should be on `main` as the most public and obvious place to find things ilke this document.

File: CONTRIBUTING.md
Comment (handrews) [inline CONTRIBUTING.md:174]: What about PRs that publish to the spec site but don't change the spec document?

File: CONTRIBUTING.md
Comment (handrews) [inline CONTRIBUTING.md:205]: I don't think we defined "headline feature" although it's fairly obvious.  Just commenting in case we want to give it a definition- I'm fine with it as-is, though.

File: style-guide.md
Comment (handrews) [inline style-guide.md:19]: I think of this as YAML notation, although now that I read this I realize that JS might allow bar key names and require quoted values, so I guess this makes sense.  But to folks like me who avoid JS like the plague, it's not the most obvious statement.  I'm not going to be upset if it's left as-is, though.

File: style-guide.md
Comment (handrews) [inline style-guide.md:20]: Can we add some note that "primitive" is not a formal type and should not be written `primitive`?  We messed this up in the patch releases in a few places.

File: style-guide.md
Comment (handrews) [inline style-guide.md:21]: We should probably provide a link for people who don't get the joke (or the grammar reference):  https://www.latimes.com/archives/blogs/jacket-copy/story/2011-06-30/goodbye-oxford-comma-hello-shatner-comma , https://www.urbandictionary.com/define.php?term=Shatner%20commas, https://elissadigital.wordpress.com/2013/09/16/your-writing-shouldnt-sound-like-william-shatner-or-say-no-to-the-shatner-comma/ should suffice for "Shatner comma" (maybe not urban dictionary- it's ads are sometimes 😬 ) and wikipedia has an [Oxford comma](https://en.wikipedia.org/wiki/Serial_comma) explanation.

File: style-guide.md
Comment (handrews) [inline style-guide.md:34]: Possibly too much of a corner case, but I do sometimes talk about Schema Object "fields" when talking about it in the context of OAS Objects as a whole, to avoid complicated "fields or keywords" phrasing.  If we want a rule about this it might be worth an issue to discuss rather than holding up this PR.

**Review by handrews — COMMENTED** [review]
Excellent work, @lornajane , thank you so much!

I have various minor comments but no dealbreakers.

File: CONTRIBUTING.md
Comment (lornajane) [reply CONTRIBUTING.md:13]: My logic here was that almost everything needs to go into 3.2 (because we don't update 3.1) and we don't want to update this doc every time so this was the most future-proof option .... I think removing "patch" totally works though!

File: CONTRIBUTING.md
Comment (lornajane) [reply CONTRIBUTING.md:174]: I don't think that is a special case, it just needs a review by TSC or maintainer. Happy to add an exception if there's a rule here that I didn't know about.

File: CONTRIBUTING.md
Comment (lornajane) [reply CONTRIBUTING.md:257]: I'll skip this, people should look in the directory and find a template. The current one has some problems and is stupidly named, so just directing people to the directory means this file won't be outdated if I ever fix that problem!

File: style-guide.md
Comment (lornajane) [reply style-guide.md:20]: I'm not sure how to provide an exhaustive list of words that shouldn't be formatted. The list above should help us just format the ones we do want, I think?

File: CONTRIBUTING.md
Comment (lornajane) [reply CONTRIBUTING.md:121]: I think we can leave this instruction as-is for now, it's not a problem if we do milestone the PRs but it's a good tip about the branches! And it can be a nice visual indicators on the PRs as well if do end up holding them open for a while.

File: style-guide.md
Comment (handrews) [reply style-guide.md:20]: Oh, I definitely wouldn't try to be exhaustive, and maybe there's a better way to handle it than a word list.

The [problem we have in the new patch releases](https://spec.openapis.org/oas/v3.1.1#style-values) is that `primitive` looks like a literal `type` value impacting the `style` field, when really it's a category encompassing the literal values `number`, `integer`, `string`, `boolean`, and `"null"`.  That's confusing, although we do use "primitives" with an [example using a `string`](https://spec.openapis.org/oas/v3.1.1#example-basic-multipart-form) and in [an example in another section](https://spec.openapis.org/oas/v3.1.1#primitive-example).

Ideally, our style guide would help us avoid these category-vs-literal-value errors (although I don't know that we have any other such errors).

File: CONTRIBUTING.md
Comment (lornajane) [reply CONTRIBUTING.md:142]: I've structured the mentions of discussions in the document something like this:
 - under "how to contribute": use discussions to give us some feedback or answer a question
 - in the discussions section, all about discussions
 - in the issues section: for big changes, use a proposal
 - in this section (pull requests): for big changes, use a proposal
 - in the proposals section: start the proposal process with a discussion

I'm quite comfortable with the structure as it is, trying to base it around a pyramid of user engagement where first they are visiting, then they are participating, then they are contributing, then they are re-engineering. I'm not wedded to it though if there's somewhere else we should mention it! I'm also not too worried if people open an issue when they should have used a discussion - engagement is good and there's nothing irreversible!

File: CONTRIBUTING.md
Comment (lornajane) [reply CONTRIBUTING.md:152]: You are right of course, but it's already the longest title in the document .... I do have the working emphasised in the next line, could that be good enough?

File: style-guide.md
Comment (lornajane) [reply style-guide.md:21]: I am not linking to urban dictionary!! But I'll edit (I wasn't familiar with the term myself)

File: CONTRIBUTING.md
Comment (lornajane) [reply CONTRIBUTING.md]: Updated, I _think_ it makes more sense now.

File: style-guide.md
Comment (lornajane) [reply style-guide.md]: OK, links added.

**Review by ralfhandl — APPROVED** [review]

**Review by handrews — APPROVED** [review]
I'm happy with this as-is, any remaining concerns are more open-ended and can be handled as follow-up discussions.

**Pull request merged**
